### PR TITLE
Restore manganis optimizations

### LINKED ISSUE #3229: Manganis 0.6.0-alpha.5 using image() in asset! results in `macro expansion ignores token , and any following`
**Problem**

When attempting to use the `asset!` macro as described in documentation [here](https://github.com/DioxusLabs/dioxus/blob/af53354b28981ff84d4e533798f1f8920371ef92/packages/manganis/manganis-macro/src/asset.rs#L26-L33) an error is thrown not allowing the usage of image() to specify image format & size of the asset.

```
20:35:26 [cargo] error: macro expansion ignores token `,` and any following
  --> src/main.rs:38:34
   |
37 |       const AUTUMN_LOGO: Asset = manganis::asset!(
   |  ________________________________-
38 | |         "/assets/autumn-logo.png",
   | |                                  ^
39 | |         asset::image().format(ImageType::Avif).size(64, 64)
40 | |     );
   | |_____- caused by the macro expansion here
   |
   = note: the usage of `manganis::asset!` is likely invalid in expression context
```

Setting the type of `AUTUMN_LOGO` to `ImageAsset` results in mismatched types error, unknown if this is even the intended solution though.
```
error[E0308]: mismatched types
  --> src/main.rs:37:37
   |
37 |       const AUTUMN_LOGO: ImageAsset = manganis::asset!(
   |  _____________________________________^
38 | |         "/assets/autumn-logo.png",
39 | |         asset::image().format(ImageType::Avif).size(64, 64)
40 | |     );
   | |_____^ expected `ImageAsset`, found `Asset`
   |
   = note: this error originates in the macro `manganis::asset` (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Steps To Reproduce**

- Use asset! macro with image() to specify image format & size
```
    const AUTUMN_LOGO: Asset = manganis::asset!(
        "/assets/autumn-logo.png",
        asset::image().format(ImageType::Avif).size(64, 64)
    );
```

This [GitHub repo](https://github.com/hyziri/dioxus_manganis_bug) is a minimal example of where this bug can be replicated.

**Environment:**
 - Dioxus version:  0.6.0-alpha.5 & also occurs in 0.6.0-alpha.4
 - Rust version:    occurs in both stable & nightly
 - OS info:         Linux
 - App platform:    occurs in both web & fullstack

**Questionnaire**

I'm interested in fixing this myself but don't know where to start

### PR DESCRIPTION
**Pull request by ealmloff**
This PR will restore the bulk of manganis with a slightly different serialization approach. It also restores all of the documentation that was removed in #2779. Instead of parsing the builder inside of the macro and serializing that into json, this PR serializes the builder directly. This should make it easier to add new asset types to manganis and make the macro significantly more flexible.

Previously the macro would only accept literals, now it will be able to accept any const type:
```rust
const ASSET: Asset = asset!("/assets/image.png");
const OPTIONS: ImageAssetOptions = ImageAssetOptions::new().with_format(ImageFormat::Avif);
const AVIF_ASSET: Asset = asset!("/assets/image.png", OPTIONS);

```

The path itself still needs to be a literal string so we can validate and hash the contents at compile time, but the rest of the settings are just const rust

closes #3229

### CODE REVIEW

File: packages/const-serialize/README.md
Comment (ealmloff) [inline packages/const-serialize/README.md:45]: Highly recommend reading this before reviewing const-serialize. Const rust is very limited which makes some of the code look odd. The serialization framework ends up looking very similar to zero copy serialization frameworks like:
- postcard-forth (a reimplementation of postcard serialization by the original author): https://onevariable.com/blog/pods-from-scratch
- zerocopy: https://github.com/google/zerocopy
Unlike zero copy serialization, const serialize normalizes endianness across platforms

**Comment by DogeDark** [comment]
I tried testing this on Windows 10 and I got this behavior:
- assets are optimized and the optimized assets are placed in my top-level assets folder (nothing about assets in the target folder)
- the link returned by manganis for use in `img` src/`document::Link` is weird/doesn't work: 
```
http://127.0.0.1:8080/assets///?\C:\Users\Darka\Documents\Projects\DioxusLabs\temp\web-manganis-test-3195\assets\main-e6ddccc1e64cc12d.css=
```

**Comment by ealmloff** [comment]
> I tried testing this on Windows 10 and I got this behavior:
> 
> * trying to use a folder asset and it errors with access denied
> * assets are optimized and the optimized assets are placed in my top-level assets folder (nothing about assets in the target folder)
> * the link returned by manganis for use in `img` src/`document::Link` is weird/doesn't work:
> 
> ```
> http://127.0.0.1:8080/assets///?\C:\Users\Darka\Documents\Projects\DioxusLabs\temp\web-manganis-test-3195\assets\main-e6ddccc1e64cc12d.css=
> ```

All of those issues should be fixed now (tested on windows 11). There was a path separator issue in manganis on windows

File: packages/cli/Cargo.toml
Comment (jkelleyrtp) [inline packages/cli/Cargo.toml:143]: it really would not be terrible to cut out the asset system (Manifest, optimizer, etc) into its own crate again... and maybe with some feature flags so we can gate `dev` mode `dx` to not have to include all this.

File: packages/cli/src/build/bundle.rs
Comment (jkelleyrtp) [inline packages/cli/src/build/bundle.rs:385]: it might be a good idea to do all this under `tokio::spawn_blocking()` now that we might be doing some serious optimizations.

**Review by jkelleyrtp — CHANGES_REQUESTED** [review]
Small tweaks:
- might be good to pull out the assets module from the CLI into its own crate 
- some of the code we generate in the macro could be moved into a const function or two
- some docs in some places refernce the old `mg!()` or are incomplete, presumably since they're already incomplete.
- it would be nice if the serialized bytes were valid json but I get how the const_serialize can't do that in a general case.
- double check any issues with us not properly cleaning the build dirs - I want to make sure there's no weirdness with leftover html/assets/gradle files still showing up between builds leaving sticky state behind.

Overall much better!

File: packages/cli/src/build/bundle.rs
Comment (ealmloff) [reply packages/cli/src/build/bundle.rs:384]: Good catch, I didn't notice that got removed when assets were removed.

File: packages/manganis-core/src/images.rs
Comment (ealmloff) [reply packages/manganis-core/src/images.rs:98]: ImageAsset doesn't exist. We can't support metadata about assets like lqip without injecting it in the linker, so I didn't add it in this PR because of the time constraint

File: packages/manganis-core/Cargo.toml
Comment (ealmloff) [reply packages/manganis-core/Cargo.toml]: It doesn't change the diff much but I pulled that change out into https://github.com/DioxusLabs/dioxus/pull/3245. Most of manganis wasn't in the repo at all.

File: packages/manganis-macro/src/asset.rs
Comment (ealmloff) [reply packages/manganis-macro/src/asset.rs]: I could inline some constants, but I think that would make things less readable the inputs/steps would be less documented.

I added some more comments about why we need each step. We are running into some limitations of const rust here. We can't return &'static str we generate in const function and into_asset_options can't easily be turned into a trait

File: packages/cli/Cargo.toml
Comment (ealmloff) [reply packages/cli/Cargo.toml:143]: Agreed. I would like to pull out asset optimization and resolution into a separate crate, but I don't want to block the release and it would take longer to make an more stable API. How about a follow up PR after more pressing issues like the base path are fixed? I have an issue ready once this is merged

**Comment by ealmloff** [comment]
> it would be nice if the serialized bytes were valid json but I get how the const_serialize can't do that in a general case.

I think that is possible, but it would require a lot more logic to parse and format strings and numbers at compile time. I'm not sure how that would effect compile times, but it could be worth exploring in the future

I don't love having a bespoke serialization format. https://github.com/jamesmunns/postcard is a much simpler well defined serialization format that might be easier to target than json. I think it is pretty similar to what we are currently generating

**Comment by DogeDark** [comment]
> All of those issues should be fixed now (tested on windows 11). There was a path separator issue in manganis on windows

I still have all the same issues on the latest commit except the folder asset now panics as todo during the CLI's asset stage

**Comment by ealmloff** [comment]
> * assets are optimized and the optimized assets are placed in my top-level assets folder (nothing about assets in the target folder)
> * the link returned by manganis for use in `img` src/`document::Link` is weird/doesn't work:

Can you share the code you used that causes those issues with the latest version? This code works as I would expect on windows:
```rust
use dioxus::prelude::*;

fn main() {
    launch(|| {
        let asset = asset!("/assets/img.png");
        rsx! {
            img {
                src: "{asset}"
            }
        }
    })
}
```

**Comment by DogeDark** [comment]
@ealmloff https://github.com/DogeDark/manganis-3195

**Comment by ealmloff** [comment]
> @ealmloff https://github.com/DogeDark/manganis-3195

Thanks, that example is working now as well. I was testing windows on desktop where path separator in the CLI is the same as the target platform path separator. WASM has unix path separators so it was stripping the wrong path when cross building to web

**Comment by DogeDark** [comment]
> Thanks, that example is working now as well.

I can confirm, it works!

**Review by jkelleyrtp — APPROVED** [review]

**Pull request merged**
